service/type.go

```diff
@@ -76,32 +76,38 @@ type Service struct {
 	Source string `hash:"name:9" validate:"required,printascii"`
 }
 
-// Dependency represents a Docker container and it holds instructions about
-// how it should run.
-type Dependency struct {
-	// Key is the key of dependency.
-	Key string `hash:"name:1" validate:"printascii"`
-
-	// Image is the Docker image.
-	Image string `hash:"name:2" validate:"printascii"`
-
+// Configuration represents the main Docker container and it holds instructions about
+// how it should run. Similar to Dependency.
+type Configuration struct {
 	// Volumes are the Docker volumes.
-	Volumes []string `hash:"name:3" validate:"unique,dive,printascii"`
+	Volumes []string `hash:"name:1" validate:"unique,dive,printascii"`
 
 	// VolumesFrom are the docker volumes-from from.
-	VolumesFrom []string `hash:"name:4" validate:"unique,dive,printascii"`
+	VolumesFrom []string `hash:"name:2" validate:"unique,dive,printascii"`
 
 	// Ports holds ports configuration for container.
-	Ports []string `hash:"name:5" validate:"unique,dive,portmap"`
+	Ports []string `hash:"name:3" validate:"unique,dive,portmap"`
 
 	// Command is the Docker command which will be executed when container started.
-	Command string `hash:"name:6" validate:"printascii"`
+	Command string `hash:"name:4" validate:"printascii"`
 
 	// Argument holds the args to pass to the Docker container
-	Args []string `hash:"name:7" validate:"dive,printascii"`
+	Args []string `hash:"name:5" validate:"dive,printascii"`
 
 	// Env is a slice of environment variables in key=value format.
-	Env []string `hash:"name:8" validate:"unique,dive,env"`
+	Env []string `hash:"name:6" validate:"unique,dive,env"`
+}
+
+// Dependency represents a Docker container and it holds instructions about
+// how it should run.
+type Dependency struct {
+	*Configuration
+
+	// Key is the key of dependency.
+	Key string `hash:"name:d1" validate:"printascii"`
+
+	// Image is the Docker image.
+	Image string `hash:"name:d2" validate:"printascii"`
```

Comment:
As it is using inheritance, I choose to name the hash with the prefix "d" so it doesn't override future properties added to Configuration struct.

Reply:
I've just tested structhash package and we don't need to prefix for hash. With struct composition when the two fields with the same name hash is different.
```go
package main

import (
    "fmt"

    "github.com/cnf/structhash"
)

type A struct {
    N string `hash:"name:1"`
    M string `hash:"name:2"`
    B
}

type B struct {
    M string `hash:"name:2"`
}

func main() {
    fmt.Printf("%x\n", structhash.Dump(A{N: "n", M: "m", B: B{M: "m"}}, 0))
    fmt.Printf("%x\n", structhash.Dump(A{N: "n", M: "m-", B: B{M: "m"}}, 0))

    fmt.Printf("%x\n", structhash.Sha1(A{N: "n", M: "m", B: B{M: "m"}}, 0))
    fmt.Printf("%x\n", structhash.Sha1(A{N: "n", M: "m-", B: B{M: "m"}}, 0))
}

```
➜  go run main.go
7b313a226e222c323a226d222c423a7b323a226d227d7d
7b313a226e222c323a226d2d222c423a7b323a226d227d7d
338eec5b413091c4db0e9aa057e861549ca45d9e
11ac72d3248a5dd00d65a1b1640a1fb6bc8326c3
```

Reply:
nice catch 👍  I removed the prefix